src/deriveNativeScriptHash.c

```diff
@@ -0,0 +1,458 @@
+
+#define ASSERT_UI_SCRIPT_TYPE_SANITY() ASSERT(ctx->ui_scriptType >= UI_SCRIPT_PUBKEY_PATH && ctx->ui_scriptType <= UI_SCRIPT_INVALID_HEREAFTER);
+#define HEADER ui_native_script_header[ctx->ui_scriptType]
+
+static void deriveScriptHash_display_ui_position(uint8_t level, ui_callback_fn_t* callback)
+{
+	ASSERT_UI_SCRIPT_TYPE_SANITY();
+	TRACE();
+
+	// For level n the longest possible position is: "x.x.x...x"
+	// where the x is 2^32-1 (length 10) and there are up to level xs
+	// + the leading "Position: "
+	char position[10 + 11 * level + 1];
+	int offset = snprintf(position, 10 + 1, "Position: ");
+	for (int i = 1; i <= level; i++) {
+		snprintf(&position[offset], 11 + 1, "%u.", ctx->complexScripts[i].totalScripts - ctx->complexScripts[i].remainingScripts + 1);
+		// snprintf returns 0, https://github.com/LedgerHQ/nanos-secure-sdk/issues/28
+		offset += strlen(&position[offset]);
+	}
+
+	// remove any trailing '.'
+	position[offset - 1] = '\0';
+
+	ui_displayPaginatedText(
+	        HEADER,
+	        position,
+	        callback
+	);
+}
+
+enum {
+	DISPLAY_UI_STEP_POSITION = 200,
+	DISPLAY_UI_STEP_SCRIPT_CONTENT,
+	DISPLAY_UI_STEP_RESPOND,
+	DISPLAY_UI_STEP_INVALID
+};
+
+static void deriveScriptHash_display_ui_runStep()
+{
+	TRACE("ui_step = %d", ctx->ui_step);
+	ASSERT_UI_SCRIPT_TYPE_SANITY();
+
+	ui_callback_fn_t* this_fn = deriveScriptHash_display_ui_runStep;
+	UI_STEP_BEGIN(ctx->ui_step, this_fn);
+
+	UI_STEP(DISPLAY_UI_STEP_POSITION) {
+		// For complex scripts we reduce the current level by 1
+		// Because they already have the level increased by 1
+		int levelOffset = ctx->ui_scriptType == UI_SCRIPT_ALL
+		                  || ctx->ui_scriptType == UI_SCRIPT_ANY
+		                  || ctx->ui_scriptType == UI_SCRIPT_N_OF_K
+		                  ? -1 : 0;
+		uint8_t level = ctx->level + levelOffset;
+		// if we are on the root level don't show the position
+		if (level == 0) {
+			TRACE("Skip showing position");
+			UI_STEP_JUMP(DISPLAY_UI_STEP_SCRIPT_CONTENT);
+		}
+		deriveScriptHash_display_ui_position(level, this_fn);
+	}
+
+	UI_STEP(DISPLAY_UI_STEP_SCRIPT_CONTENT) {
+		TRACE("ui_scriptType = %d", ctx->ui_scriptType);
+		switch (ctx->ui_scriptType) {
+		case UI_SCRIPT_PUBKEY_PATH:
+			ui_displayPathScreen(
+			        HEADER,
+			        &ctx->scriptContent.pubkeyPath,
+			        this_fn
+			);
+			break;
+		case UI_SCRIPT_PUBKEY_HASH:
+			ui_displayBech32Screen(
+			        HEADER,
+			        "addr_vkh",
+			        ctx->scriptContent.pubkeyHash,
+			        ADDRESS_KEY_HASH_LENGTH,
+			        this_fn
+			);
+			break;
+		case UI_SCRIPT_ALL:
+		case UI_SCRIPT_ANY: {
+			char text[35];
+			SAFE_SPRINTF(text, "Contains %d nested scripts", ctx->complexScripts[ctx->level].remainingScripts);
+
+			ui_displayPaginatedText(
+			        HEADER,
+			        text,
+			        this_fn
+			);
+		}
+		break;
```

Comment:
The location of this `break` looks suspicious --- shouldn't it go inside {}?. Also, it is important to be consistent in using {} for cases consistently --- either not at all, or for each case in a switch. I guess we should use {} whenever the cases are longer and involve indentation, as is the case here.